File: articles.py
"""
NewsHub Cloud — Netlify Function  /.netlify/functions/articles
(accessible as /api/articles via the redirect in netlify.toml)

Fetches all RSS/Atom feeds server-side and returns a sorted JSON array.
Response is CDN-cached for 30 minutes with stale-while-revalidate so
users always get an instant response after the first visit.

All Python 3 stdlib only — no pip install required.
"""

import json, ssl, re, urllib.request, urllib.error
import xml.etree.ElementTree as ET
from datetime import datetime
from concurrent.futures import ThreadPoolExecutor, as_completed, TimeoutError

# ── Feed list ────────────────────────────────────────────────
# Broken feeds replaced:
#   TLDR AI (404)        → The Batch by DeepLearning.AI
#   Scientific Am (SSL)  → New Scientist
#   Investopedia (404)   → Yahoo Finance

DEFAULT_FEEDS = [
    # ── AI & ML
    {'id':'one-useful-thing','name':'One Useful Thing',   'url':'https://www.oneusefulthing.org/feed',                              'cat':'ai'},
    {'id':'simon-willison',  'name':'Simon Willison',     'url':'https://simonwillison.net/atom/everything/',                       'cat':'ai'},
    {'id':'hf',              'name':'Hugging Face',       'url':'https://huggingface.co/blog/feed.xml',                             'cat':'ai'},
    {'id':'the-batch',       'name':'The Batch (DL.AI)',  'url':'https://www.deeplearning.ai/the-batch/feed/',                      'cat':'ai'},
    {'id':'ai-news',         'name':'AI News',            'url':'https://www.artificialintelligence-news.com/feed/',                 'cat':'ai'},
    {'id':'mit-ai',          'name':'MIT AI News',        'url':'https://news.mit.edu/rss/topic/artificial-intelligence2',          'cat':'ai'},
    {'id':'openai',          'name':'OpenAI Blog',        'url':'https://openai.com/news/rss.xml',                                  'cat':'ai'},
    {'id':'deepmind',        'name':'Google DeepMind',    'url':'https://deepmind.google/blog/rss.xml',                             'cat':'ai'},
    {'id':'import-ai',       'name':'Import AI',          'url':'https://jack-clark.net/feed/',                                     'cat':'ai'},
    # ── Tech
    {'id':'hn',              'name':'Hacker News',        'url':'https://hnrss.org/frontpage',                                      'cat':'tech'},
    {'id':'tc',              'name':'TechCrunch',         'url':'https://techcrunch.com/feed/',                                     'cat':'tech'},
    {'id':'verge',           'name':'The Verge',          'url':'https://www.theverge.com/rss/index.xml',                           'cat':'tech'},
    {'id':'ars',             'name':'Ars Technica',       'url':'https://feeds.arstechnica.com/arstechnica/index',                  'cat':'tech'},
    {'id':'engadget',        'name':'Engadget',           'url':'https://www.engadget.com/rss.xml',                                 'cat':'tech'},
    {'id':'zdnet',           'name':'ZDNet',              'url':'https://www.zdnet.com/news/rss.xml',                               'cat':'tech'},
    {'id':'tldr-tech',       'name':'TLDR Tech',          'url':'https://tldr.tech/rss',                                            'cat':'tech'},
    # ── Finance
    {'id':'cnbc',            'name':'CNBC Markets',       'url':'https://www.cnbc.com/id/100003114/device/rss/rss.html',            'cat':'finance'},
    {'id':'mwatch',          'name':'MarketWatch',        'url':'https://feeds.marketwatch.com/marketwatch/topstories/',            'cat':'finance'},
    {'id':'coindesk',        'name':'CoinDesk',           'url':'https://www.coindesk.com/arc/outboundfeeds/rss/',                  'cat':'finance'},
    {'id':'yahoo-finance',   'name':'Yahoo Finance',      'url':'https://finance.yahoo.com/news/rssindex',                         'cat':'finance'},
    {'id':'fool',            'name':'Motley Fool',        'url':'https://www.fool.com/feeds/index.aspx',                            'cat':'finance'},
    # ── General
    {'id':'bbc',             'name':'BBC News',           'url':'https://feeds.bbci.co.uk/news/rss.xml',                            'cat':'general'},
    {'id':'guardian',        'name':'The Guardian Tech',  'url':'https://www.theguardian.com/technology/rss',                       'cat':'general'},
    {'id':'mit-rev',         'name':'MIT Tech Review',    'url':'https://www.technologyreview.com/feed/',                           'cat':'general'},
    {'id':'nyt-tech',        'name':'NYT Technology',     'url':'https://rss.nytimes.com/services/xml/rss/nyt/Technology.xml',      'cat':'general'},
    {'id':'new-scientist',   'name':'New Scientist',      'url':'https://www.newscientist.com/feed/home/',                          'cat':'general'},
]

HEADERS = {
    'User-Agent':      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 NewsHub/3.0',
    'Accept':          'application/rss+xml, application/atom+xml, application/xml, text/xml, */*',
    'Accept-Language': 'en-US,en;q=0.9',
    'Cache-Control':   'no-cache',
}

FEED_TIMEOUT = 7   # seconds per feed — leaves headroom within Netlify's 10s limit

# ── Helpers ──────────────────────────────────────────────────
def strip_tags(html):
    if not html: return ''
    s = re.sub(r'<[^>]+>', ' ', str(html))
    for code, char in [('&amp;','&'),('&lt;','<'),('&gt;','>'),
                       ('&quot;','"'),('&apos;',"'"),('&#039;',"'"),('&nbsp;',' ')]:
        s = s.replace(code, char)
    return re.sub(r'\s+', ' ', s).strip()


def parse_date(s):
    if not s: return None
    s = s.strip()
    try:
        from email.utils import parsedate_to_datetime
        return parsedate_to_datetime(s).isoformat()
    except Exception: pass
    try:
        return datetime.fromisoformat(re.sub(r'Z$', '+00:00', s)).isoformat()
    except Exception: pass
    return None


def get_text(element, tag):
    """Get text from an XML element — explicit is-None check avoids DeprecationWarning."""
    el = element.find(tag)
    if el is None:
        el = element.find(f'{{http://www.w3.org/2005/Atom}}{tag}')
    return el.text if el is not None else None


def parse_feed(raw_bytes, feed):
    text = raw_bytes.decode('utf-8', errors='replace')
    text = re.sub(r'\sxmlns(?::[^=]+)?="[^"]*"', '', text)
    text = re.sub(r'[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]', '', text)
    try:
        root = ET.fromstring(text)
    except ET.ParseError:
        return []

    is_atom = 'feed' in root.tag.lower()
    articles = []

    if is_atom:
        entries = (root.findall('entry') or
                   root.findall('{http://www.w3.org/2005/Atom}entry'))
        for e in entries[:40]:
            title = strip_tags(get_text(e, 'title')) or 'Untitled'
            link = ''
            for lel in (e.findall('link') + e.findall('{http://www.w3.org/2005/Atom}link')):
                rel  = lel.get('rel', 'alternate')
                href = lel.get('href', '') or (lel.text or '').strip()
                if rel == 'alternate' or (not link and href):
                    link = href.strip(); break
            date = parse_date(get_text(e, 'published') or get_text(e, 'updated'))
            desc = strip_tags(get_text(e, 'summary') or get_text(e, 'content'))[:280]
            if link:
                articles.append({'title':title,'link':link,'date':date,'desc':desc,
                                  'source':feed['name'],'cat':feed['cat'],'feedId':feed['id']})
    else:
        for it in root.findall('.//item')[:40]:
            title = strip_tags(get_text(it, 'title')) or 'Untitled'
            link  = (get_text(it, 'link') or get_text(it, 'guid') or '').strip()
            date  = parse_date(get_text(it, 'pubDate') or get_text(it, 'date') or get_text(it, 'updated'))
            desc  = strip_tags(
                get_text(it, 'description') or get_text(it, 'summary') or
                get_text(it, '{http://purl.org/rss/1.0/modules/content/}encoded') or '')[:280]
            if link:
                articles.append({'title':title,'link':link,'date':date,'desc':desc,
                                  'source':feed['name'],'cat':feed['cat'],'feedId':feed['id']})
    return articles


def fetch_one(feed):
    """Fetch a single feed. Returns list of articles (empty on any failure)."""
    try:
        req = urllib.request.Request(feed['url'], headers=HEADERS)
        try:
            with urllib.request.urlopen(req, timeout=FEED_TIMEOUT) as resp:
                raw = resp.read()
        except ssl.SSLError:
            # SSL fallback — retries without certificate verification
            ctx = ssl.create_default_context()
            ctx.check_hostname = False
            ctx.verify_mode = ssl.CERT_NONE
            with urllib.request.urlopen(req, timeout=FEED_TIMEOUT, context=ctx) as resp:
                raw = resp.read()
        return parse_feed(raw, feed)
    except Exception:
        return []


# ── Netlify handler ──────────────────────────────────────────
def handler(event, context):
    # Handle CORS preflight
    if event.get('httpMethod') == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Methods': 'GET, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type',
            },
            'body': '',
        }

    articles = []

    # All feeds start simultaneously — collect results up to 9 seconds
    with ThreadPoolExecutor(max_workers=len(DEFAULT_FEEDS)) as ex:
        futures = {ex.submit(fetch_one, f): f for f in DEFAULT_FEEDS}
        try:
            for fut in as_completed(futures, timeout=9):
                articles.extend(fut.result())
        except TimeoutError:
            # Return whatever completed before the cutoff
            for fut in futures:
                if fut.done():
                    try: articles.extend(fut.result())
                    except Exception: pass

    articles.sort(key=lambda a: a.get('date') or '', reverse=True)

    return {
        'statusCode': 200,
        'headers': {
            'Content-Type':                'application/json; charset=utf-8',
            # Fresh 30 min; serve stale while revalidating for up to 24 h
            'Cache-Control':               'public, s-maxage=1800, stale-while-revalidate=86400',
            'Access-Control-Allow-Origin': '*',
        },
        'body': json.dumps(articles, ensure_ascii=False),
    }

File: netlify/functions/articles.js
const https = require('https');
const http = require('http');
const { URL } = require('url');

const DEFAULT_FEEDS = [
  {id:'one-useful-thing',name:'One Useful Thing',url:'https://www.oneusefulthing.org/feed',cat:'ai'},
  {id:'simon-willison',name:'Simon Willison',url:'https://simonwillison.net/atom/everything/',cat:'ai'},
  {id:'hf',name:'Hugging Face',url:'https://huggingface.co/blog/feed.xml',cat:'ai'},
  {id:'the-batch',name:'The Batch (DL.AI)',url:'https://www.deeplearning.ai/the-batch/feed/',cat:'ai'},
  {id:'ai-news',name:'AI News',url:'https://www.artificialintelligence-news.com/feed/',cat:'ai'},
  {id:'mit-ai',name:'MIT AI News',url:'https://news.mit.edu/rss/topic/artificial-intelligence2',cat:'ai'},
  {id:'openai',name:'OpenAI Blog',url:'https://openai.com/news/rss.xml',cat:'ai'},
  {id:'deepmind',name:'Google DeepMind',url:'https://deepmind.google/blog/rss.xml',cat:'ai'},
  {id:'import-ai',name:'Import AI',url:'https://jack-clark.net/feed/',cat:'ai'},
  {id:'google-ai',name:'Google AI Blog',url:'https://blog.google/technology/ai/rss/',cat:'ai'},
  {id:'nvidia-dev',name:'Nvidia AI Blog',url:'https://developer.nvidia.com/blog/feed/',cat:'ai'},
  {id:'arxiv-ai',name:'ArXiv CS.AI',url:'https://rss.arxiv.org/rss/cs.AI',cat:'ai'},
  {id:'lesswrong',name:'LessWrong',url:'https://www.lesswrong.com/feed.xml',cat:'ai'},
  {id:'the-gradient',name:'The Gradient',url:'https://thegradient.pub/rss/',cat:'ai'},
  {id:'towards-ds',name:'Towards Data Science',url:'https://towardsdatascience.com/feed',cat:'ai'},
  {id:'towards-ai',name:'Towards AI',url:'https://pub.towardsai.net/feed',cat:'ai'},
  {id:'anthropic',name:'Anthropic',url:'https://news.google.com/rss/search?q=site:anthropic.com&hl=en-US&gl=US&ceid=US:en',cat:'ai'},
  {id:'meta-ai',name:'Meta AI',url:'https://news.google.com/rss/search?q=site:ai.meta.com&hl=en-US&gl=US&ceid=US:en',cat:'ai'},
  {id:'mistral',name:'Mistral AI',url:'https://news.google.com/rss/search?q=site:mistral.ai&hl=en-US&gl=US&ceid=US:en',cat:'ai'},
  {id:'a16z-ai',name:'a16z AI',url:'https://news.google.com/rss/search?q=site:a16z.com+AI&hl=en-US&gl=US&ceid=US:en',cat:'ai'},
  {id:'perplexity',name:'Perplexity AI',url:'https://news.google.com/rss/search?q=site:perplexity.ai&hl=en-US&gl=US&ceid=US:en',cat:'ai'},
  {id:'superhuman-ai',name:'Superhuman AI',url:'https://news.google.com/rss/search?q=superhuman+AI+newsletter&hl=en-US&gl=US&ceid=US:en',cat:'ai'},
  {id:'venturebeat-ai',name:'VentureBeat AI',url:'https://news.google.com/rss/search?q=site:venturebeat.com+AI&hl=en-US&gl=US&ceid=US:en',cat:'ai'},
  {id:'investopedia',name:'Investopedia',url:'https://news.google.com/rss/search?q=site:investopedia.com&hl=en-US&gl=US&ceid=US:en',cat:'finance'},
  {id:'barrons',name:"Barron's",url:'https://news.google.com/rss/search?q=site:barrons.com&hl=en-US&gl=US&ceid=US:en',cat:'finance'},
  {id:'rundown-ai',name:'The Rundown AI',url:'https://therundown.substack.com/feed',cat:'ai'},
  {id:'bens-bites',name:"Ben's Bites",url:'https://www.bensbites.com/feed',cat:'ai'},
  {id:'tldr-ai',name:'TLDR AI',url:'https://tldr.tech/api/rss/ai',cat:'ai'},
  {id:'hn',name:'Hacker News',url:'https://hnrss.org/frontpage',cat:'tech'},
  {id:'tc',name:'TechCrunch',url:'https://techcrunch.com/feed/',cat:'tech'},
  {id:'verge',name:'The Verge',url:'https://www.theverge.com/rss/index.xml',cat:'tech'},
  {id:'ars',name:'Ars Technica',url:'https://feeds.arstechnica.com/arstechnica/index',cat:'tech'},
  {id:'engadget',name:'Engadget',url:'https://www.engadget.com/rss.xml',cat:'tech'},
  {id:'zdnet',name:'ZDNet',url:'https://www.zdnet.com/news/rss.xml',cat:'tech'},
  {id:'tldr-tech',name:'TLDR Tech',url:'https://tldr.tech/rss',cat:'tech'},
  {id:'wired',name:'Wired',url:'https://www.wired.com/feed/rss',cat:'tech'},
  {id:'ieee',name:'IEEE Spectrum',url:'https://spectrum.ieee.org/feeds/feed.rss',cat:'tech'},
  {id:'9to5mac',name:'9to5Mac',url:'https://9to5mac.com/feed/',cat:'tech'},
  {id:'fast-company',name:'Fast Company',url:'https://www.fastcompany.com/latest/rss',cat:'tech'},
  {id:'the-information',name:'The Information',url:'https://www.theinformation.com/feed',cat:'tech'},
  {id:'wsj-tech',name:'WSJ Tech',url:'https://feeds.a.dj.com/rss/RSSWSJD.xml',cat:'tech'},
  {id:'cnbc',name:'CNBC Markets',url:'https://www.cnbc.com/id/100003114/device/rss/rss.html',cat:'finance'},
  {id:'mwatch',name:'MarketWatch',url:'https://feeds.marketwatch.com/marketwatch/topstories/',cat:'finance'},
  {id:'coindesk',name:'CoinDesk',url:'https://www.coindesk.com/arc/outboundfeeds/rss/',cat:'finance'},
  {id:'yahoo-finance',name:'Yahoo Finance',url:'https://finance.yahoo.com/news/rssindex',cat:'finance'},
  {id:'fool',name:'Motley Fool',url:'https://www.fool.com/feeds/index.aspx',cat:'finance'},
  {id:'bloomberg-tech',name:'Bloomberg Tech',url:'https://feeds.bloomberg.com/technology/news.rss',cat:'finance'},
  {id:'bloomberg-mkts',name:'Bloomberg Markets',url:'https://feeds.bloomberg.com/markets/news.rss',cat:'finance'},
  {id:'wsj-markets',name:'WSJ Markets',url:'https://feeds.a.dj.com/rss/RSSMarketsMain.xml',cat:'finance'},
  {id:'ft',name:'Financial Times',url:'https://www.ft.com/rss/home/uk',cat:'finance'},
  {id:'seeking-alpha',name:'Seeking Alpha',url:'https://seekingalpha.com/feed.xml',cat:'finance'},
  {id:'economist',name:'The Economist',url:'https://www.economist.com/finance-and-economics/rss.xml',cat:'finance'},
  {id:'bbc',name:'BBC News',url:'https://feeds.bbci.co.uk/news/rss.xml',cat:'general'},
  {id:'guardian',name:'The Guardian Tech',url:'https://www.theguardian.com/technology/rss',cat:'general'},
  {id:'mit-rev',name:'MIT Tech Review',url:'https://www.technologyreview.com/feed/',cat:'general'},
  {id:'nyt-tech',name:'NYT Technology',url:'https://rss.nytimes.com/services/xml/rss/nyt/Technology.xml',cat:'general'},
  {id:'new-scientist',name:'New Scientist',url:'https://www.newscientist.com/feed/home/',cat:'general'},
  {id:'npr',name:'NPR News',url:'https://feeds.npr.org/1001/rss.xml',cat:'breaking'},
  {id:'al-jazeera',name:'Al Jazeera',url:'https://www.aljazeera.com/xml/rss/all.xml',cat:'breaking'},
  {id:'guardian-world',name:'Guardian World',url:'https://www.theguardian.com/world/rss',cat:'breaking'},
  {id:'reuters',name:'Reuters',url:'https://news.google.com/rss/search?q=site:reuters.com&hl=en-US&gl=US&ceid=US:en',cat:'breaking'},
  {id:'sky-news',name:'Sky News',url:'https://feeds.skynews.com/feeds/rss/world.xml',cat:'breaking'},
  // Sports
  {id:'espn',name:'ESPN Top Stories',url:'https://www.espn.com/espn/rss/news',cat:'sports'},
  {id:'espn-nba',name:'ESPN NBA',url:'https://www.espn.com/espn/rss/nba/news',cat:'sports'},
  {id:'espn-nfl',name:'ESPN NFL',url:'https://www.espn.com/espn/rss/nfl/news',cat:'sports'},
  {id:'espn-soccer',name:'ESPN Soccer',url:'https://www.espn.com/espn/rss/soccer/news',cat:'sports'},
  {id:'bbc-sport',name:'BBC Sport',url:'https://feeds.bbci.co.uk/sport/rss.xml',cat:'sports'},
  {id:'bbc-football',name:'BBC Sport Football',url:'https://feeds.bbci.co.uk/sport/football/rss.xml',cat:'sports'},
  {id:'sky-sports',name:'Sky Sports',url:'https://www.skysports.com/rss/12040',cat:'sports'},
  {id:'fotmob',name:'FotMob',url:'https://news.google.com/rss/search?q=fotmob+football+news',cat:'sports'},
  {id:'athletic',name:'The Athletic',url:'https://news.google.com/rss/search?q=site:theathletic.com',cat:'sports'},
  {id:'bleacher-report',name:'Bleacher Report',url:'https://news.google.com/rss/search?q=site:bleacherreport.com',cat:'sports'},
  {id:'sports-illustrated',name:'Sports Illustrated',url:'https://news.google.com/rss/search?q=site:si.com',cat:'sports'},
  {id:'nba-official',name:'NBA.com News',url:'https://news.google.com/rss/search?q=site:nba.com',cat:'sports'},
  {id:'nfl-official',name:'NFL.com News',url:'https://news.google.com/rss/search?q=site:nfl.com',cat:'sports'},
];

function stripTags(html) {
  if (!html) return '';
  return String(html)
    .replace(/<[^>]+>/g,' ')
    .replace(/&amp;/g,'&').replace(/&lt;/g,'<').replace(/&gt;/g,'>')
    .replace(/&quot;/g,'"').replace(/&apos;/g,"'").replace(/&nbsp;/g,' ')
    .replace(/\s+/g,' ').trim();
}

function parseDate(s) {
  if (!s) return null;
  try { const d = new Date(s); return isNaN(d) ? null : d.toISOString(); } catch { return null; }
}

function getTag(xml, tag) {
  const patterns = [
    new RegExp('<' + tag + '[^>]*>([\\s\\S]*?)<\/' + tag + '>', 'i'),
    new RegExp('<[a-z]+:' + tag + '[^>]*>([\\s\\S]*?)</[a-z]+:' + tag + '>', 'i'),
  ];
  for (const p of patterns) {
    const m = xml.match(p);
    if (m) return m[1].replace(/<!\[CDATA\[([\s\S]*?)\]\]>/g, '$1').trim();
  }
  return '';
}

function parseFeed(text, feed) {
  const articles = [];
  const isAtom = /<feed[\s>]/i.test(text);
  if (isAtom) {
    const entries = text.match(/<entry[\s>][\s\S]*?<\/entry>/gi) || [];
    for (const e of entries.slice(0, 40)) {
      const title = stripTags(getTag(e, 'title')) || 'Untitled';
      const lm = e.match(/<link[^>]+href="([^"]+)"/i);
      const link = lm ? lm[1] : '';
      const date = parseDate(getTag(e,'published') || getTag(e,'updated'));
      const desc = stripTags(getTag(e,'summary') || getTag(e,'content')).slice(0,280);
      if (link) articles.push({title,link,date,desc,source:feed.name,cat:feed.cat,feedId:feed.id});
    }
  } else {
    const items = text.match(/<item[\s>][\s\S]*?<\/item>/gi) || [];
    for (const it of items.slice(0, 40)) {
      const title = stripTags(getTag(it,'title')) || 'Untitled';
      const link = (getTag(it,'link') || getTag(it,'guid') || '').trim();
      const date = parseDate(getTag(it,'pubDate') || getTag(it,'date'));
      const desc = stripTags(getTag(it,'description') || getTag(it,'summary')).slice(0,280);
      if (link) articles.push({title,link,date,desc,source:feed.name,cat:feed.cat,feedId:feed.id});
    }
  }
  return articles;
}

function fetchFeed(feed, redirects = 0) {
  return new Promise((resolve) => {
    if (redirects > 3) return resolve([]);
    try {
      const u = new URL(feed.url);
      const lib = u.protocol === 'https:' ? https : http;
      const req = lib.request({
        hostname: u.hostname, path: u.pathname + u.search, port: u.port || undefined,
        headers: {'User-Agent':'Mozilla/5.0 NewsHub/3.0','Accept':'application/rss+xml,application/atom+xml,*/*'},
        timeout: 7000,
      }, (res) => {
        if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
          return fetchFeed({...feed, url: new URL(res.headers.location, feed.url).href}, redirects+1).then(resolve);
        }
        let data = '';
        res.setEncoding('utf8');
        res.on('data', c => data += c);
        res.on('end', () => { try { resolve(parseFeed(data, feed)); } catch { resolve([]); }});
        res.on('error', () => resolve([]));
      });
      req.on('timeout', () => { req.destroy(); resolve([]); });
      req.on('error', () => resolve([]));
      req.end();
    } catch { resolve([]); }
  });
}

const CORS = {'Access-Control-Allow-Origin':'*','Access-Control-Allow-Methods':'GET,OPTIONS','Access-Control-Allow-Headers':'Content-Type'};

exports.handler = async (event) => {
  if (event.httpMethod === 'OPTIONS') return {statusCode:200, headers:CORS, body:''};
  const timeout = (ms) => new Promise(r => setTimeout(() => r([]), ms));
  const results = await Promise.allSettled(
    DEFAULT_FEEDS.map(f => Promise.race([fetchFeed(f), timeout(8500)]))
  );
  const articles = results.flatMap(r => r.status === 'fulfilled' ? (r.value || []) : []);
  articles.sort((a,b) => (b.date||'') > (a.date||'') ? 1 : -1);
  return {
    statusCode: 200,
    headers: {...CORS, 'Content-Type':'application/json; charset=utf-8', 'Cache-Control':'public, s-maxage=1800, stale-while-revalidate=86400'},
    body: JSON.stringify(articles),
  };
};
